src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -13,193 +12,208 @@ class InMemorySagaPersister : ISagaPersister
+               }
+           });
+
             return TaskEx.CompletedTask;
         }
 
         public Task Update(IContainSagaData sagaData, SynchronizedStorageSession session, ContextBag context)
         {
-            var inMemSession = (InMemorySynchronizedStorageSession) session;
-            inMemSession.Enlist(() =>
-            {
-                data.AddOrUpdate(sagaData.Id,
-                    id => new VersionedSagaEntity(sagaData, $"{sagaData.GetType().FullName}.None.None"), // we can never end up here.
-                    (id, original) => new VersionedSagaEntity(sagaData, original.LockTokenKey, original));
-            });
+            ((InMemorySynchronizedStorageSession)session).Enlist(() =>
+           {
+               var entry = GetEntry(sagaData, context);
+
+               if (sagas.TryUpdate(sagaData.Id, entry.UpdateTo(sagaData), entry) == false)
+               {
+                   throw new Exception($"InMemorySagaPersister concurrency violation: saga entity Id[{sagaData.Id}] already saved.");
+               }
+           });
+
             return TaskEx.CompletedTask;
         }
 
-        void ValidateUniqueProperties(SagaCorrelationProperty correlationProperty, IContainSagaData saga)
+        static Entry GetEntry(IContainSagaData sagaData, ContextBag context)
         {
-            var sagaType = saga.GetType();
-            var existingSagas = new List<VersionedSagaEntity>();
-            // ReSharper disable once LoopCanBeConvertedToQuery
-            foreach (var s in data)
+            Entry entry;
+            if (context.TryGet(sagaData.Id.ToString(), out entry) == false)
             {
-                if (s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id))
-                {
-                    existingSagas.Add(s.Value);
-                }
-            }
-            var uniqueProperty = sagaType.GetProperty(correlationProperty.Name);
-
-            if (correlationProperty.Value == null)
-            {
-                var message = $"Cannot store saga with id '{saga.Id}' since the unique property '{uniqueProperty.Name}' has a null value.";
-                throw new InvalidOperationException(message);
+                throw new Exception("The saga should be retrieved with Get method before it's updated");
             }
+            return entry;
+        }
 
-            foreach (var storedSaga in existingSagas)
-            {
-                var storedSagaPropertyValue = uniqueProperty.GetValue(storedSaga.SagaData, null);
-                if (Equals(correlationProperty.Value, storedSagaPropertyValue))
-                {
-                    var message = $"Cannot store a saga. The saga with id '{storedSaga.SagaData.Id}' already has property '{uniqueProperty.Name}'.";
-                    throw new InvalidOperationException(message);
-                }
-            }
+        static CorrelationId GetKey(Type sagaType, SagaCorrelationProperty correlationProperty)
```

Comment:
I think we should inline this (or at least rename it to something like `CreateCorrelationId` to avoid similarities with Get as a lookup mechanism (like in the context bag)